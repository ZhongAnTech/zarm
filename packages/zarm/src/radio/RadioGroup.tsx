import React, {
  HTMLAttributes,
  PureComponent,
  cloneElement,
  ReactNode,
  isValidElement,
  ChangeEvent,
} from 'react';
import classnames from 'classnames';
import { BaseRadioGroupProps, RadioValue } from './PropsType';
import { Nullable } from '../utils/utilityTypes';

const getChildChecked = (children: ReactNode): Nullable<RadioValue> => {
  let checkedValue = null;
  React.Children.forEach(children, (element: ReactNode) => {
    if (isValidElement(element) && element.props && element.props.checked) {
      checkedValue = element.props.value;
    }
  });
  return checkedValue;
};

const getValue = (
  props: RadioGroup['props'],
  defaultValue: Nullable<RadioValue> = null,
): Nullable<RadioValue> => {
  if (typeof props.value !== 'undefined') {
    return props.value;
  }
  if (typeof props.defaultValue !== 'undefined') {
    return props.defaultValue;
  }
  if (getChildChecked(props.children)) {
    return getChildChecked(props.children);
  }
  return defaultValue;
};

export interface RadioGroupProps
  extends Omit<HTMLAttributes<HTMLDivElement>, 'defaultValue' | 'value' | 'onChange'>,
    BaseRadioGroupProps {
  prefixCls?: string;
}

export interface RadioGroupStates {
  value?: Nullable<RadioValue>;
}

export default class RadioGroup extends PureComponent<RadioGroupProps, RadioGroupStates> {
  static displayName = 'RadioGroup';

  static defaultProps = {
    prefixCls: 'za-radio-group',
    shape: 'radius',
    block: false,
    disabled: false,
    compact: false,
    ghost: false,
    size: 'xs',
  };

  state: RadioGroupStates = {
    value: getValue(this.props),
  };

  static getDerivedStateFromProps(nextProps: RadioGroup['props']) {
    if ('value' in nextProps) {
      return {
        value: nextProps.value,
      };
    }

    return null;
  }

  onChildChange = (value: string | number) => {
    this.setState({ value });
    const { onChange } = this.props;
    if (typeof onChange === 'function') {
      onChange(value);
    }
  };

  render() {
    const {
      prefixCls,
      className,
      size,
      shape,
      type,
      block,
      disabled,
      compact,
      ghost,
      children,
      onChange,
      defaultValue,
      value,
      ...rest
    } = this.props;
    const { value: valueState } = this.state;

    const items = React.Children.map(children, (element: React.ReactElement, index) => {
      return cloneElement(element, {
        key: index,
        type,
        shape,
        disabled: disabled || !!element.props.disabled,
        checked: valueState === element.props.value,
        onChange: (e: ChangeEvent<HTMLInputElement>) => {
          typeof element.props.onChange === 'function' && element.props.onChange(e);
          this.onChildChange(element.props.value);
        },
      });
    });

    const cls = classnames(prefixCls, className, {
      [`${prefixCls}--${type}`]: !!type,
      [`${prefixCls}--${size}`]: !!size,
      [`${prefixCls}--${shape}`]: !!shape,
      [`${prefixCls}--block`]: block,
      [`${prefixCls}--disabled`]: disabled,
      [`${prefixCls}--compact`]: compact,
      [`${prefixCls}--ghost`]: ghost,
    });

    return (
      <div className={cls} {...rest}>
        <div className={`${prefixCls}__inner`}>{items}</div>
      </div>
    );
  }
}
