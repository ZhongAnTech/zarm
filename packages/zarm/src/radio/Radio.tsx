import React, {
  PureComponent,
  ChangeEvent,
  InputHTMLAttributes,
  HTMLAttributes,
  ButtonHTMLAttributes,
} from 'react';
import classnames from 'classnames';
import { BaseRadioProps } from './PropsType';
import RadioGroup from './RadioGroup';
import Cell from '../cell';

const getChecked = (props: RadioProps, defaultChecked: boolean) => {
  if (typeof props.checked !== 'undefined') {
    return props.checked;
  }
  if (typeof props.defaultChecked !== 'undefined') {
    return props.defaultChecked;
  }
  return defaultChecked;
};

type RadioSpanProps = Omit<
  InputHTMLAttributes<HTMLInputElement>,
  'type' | 'defaultChecked' | 'checked' | 'value' | 'onChange'
>;
type RadioCellProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  'type' | 'defaultChecked' | 'checked' | 'value' | 'onChange'
>;
type RadioButtonProps = Omit<
  ButtonHTMLAttributes<HTMLButtonElement>,
  'type' | 'defaultChecked' | 'checked' | 'value' | 'onChange'
>;

export type RadioProps = Partial<RadioSpanProps & RadioCellProps & RadioButtonProps> &
  BaseRadioProps & {
    prefixCls?: string;
  };

export interface RadioStates {
  checked: boolean;
}

export default class Radio extends PureComponent<RadioProps, RadioStates> {
  static Group: typeof RadioGroup;

  static defaultProps: RadioProps = {
    prefixCls: 'za-radio',
    disabled: false,
    shape: 'radius',
  };

  state: RadioStates = {
    checked: getChecked(this.props, false),
  };

  static getDerivedStateFromProps(nextProps: RadioProps, state) {
    if ('checked' in nextProps && nextProps.checked !== state.prevChecked) {
      return {
        checked: nextProps.checked,
        prevChecked: nextProps.checked,
      };
    }

    return null;
  }

  onValueChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { disabled, onChange } = this.props;
    const { checked } = this.state;

    if (disabled) {
      return;
    }

    const newChecked = !checked;
    if (!('checked' in this.props)) {
      this.setState({ checked: newChecked });
    }

    if (typeof onChange === 'function') {
      onChange(e);
    }
  };

  render() {
    const {
      prefixCls,
      className,
      type,
      shape,
      value,
      checked,
      defaultChecked,
      disabled,
      id,
      children,
      onChange,
      ...rest
    } = this.props;
    const { checked: checkedState } = this.state;

    const cls = classnames(prefixCls, className, {
      [`${prefixCls}--checked`]: checkedState,
      [`${prefixCls}--disabled`]: disabled,
      [`${prefixCls}--untext`]: !children,
    });

    const inputRender = (
      <input
        id={id}
        type="radio"
        className={`${prefixCls}__input`}
        value={value}
        disabled={disabled}
        checked={checkedState}
        onChange={this.onValueChange}
      />
    );

    const radioRender = (
      <span className={cls} {...(rest as RadioSpanProps)}>
        <span className={`${prefixCls}__widget`}>
          <span className={`${prefixCls}__inner`} />
        </span>
        {children && <span className={`${prefixCls}__text`}>{children}</span>}
        {inputRender}
      </span>
    );

    if (type === 'cell') {
      return (
        <Cell
          disabled={disabled}
          className={className}
          onClick={() => {}}
          {...(rest as RadioCellProps)}
        >
          {radioRender}
        </Cell>
      );
    }

    if (type === 'button') {
      return (
        <button type="button" disabled={disabled} className={cls} {...(rest as RadioButtonProps)}>
          {children}
          {inputRender}
        </button>
      );
    }

    return radioRender;
  }
}
