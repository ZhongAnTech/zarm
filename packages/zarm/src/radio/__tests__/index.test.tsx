import React, { ChangeEvent } from 'react';
import { render, mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import RadioGroup from '../RadioGroup';
import Radio from '../index';

class TestRadio extends React.Component<{
  checked?: boolean;
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
  value: number;
}> {
  render() {
    return <input {...this.props} />;
  }
}

describe('Radio', () => {
  it('renders correctly', () => {
    const wrapper = render(<Radio value="0">选项一</Radio>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('defaultChecked', () => {
    const wrapper = render(
      <Radio defaultChecked value="0">
        选项一
      </Radio>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is button', () => {
    const wrapper = render(
      <Radio type="button" value="0">
        选项一
      </Radio>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is cell', () => {
    const wrapper = render(
      <Radio type="cell" value="0">
        选项一
      </Radio>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('receive new checked', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Radio value="0" onChange={onChange}>
        选项一
      </Radio>,
    );
    wrapper.setProps({ checked: true });
    wrapper.find('input[type="radio"]').simulate('change');
    expect(onChange).toBeCalled();
  });

  it('disabled', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Radio value="0" onChange={onChange}>
        选项一
      </Radio>,
    );
    wrapper.setProps({ disabled: true });
    wrapper.find('input[type="radio"]').simulate('change');
  });
});

describe('Radio.Group', () => {
  describe('#getDerivedStateFromProps', () => {
    it('should get derived state from props correctly', () => {
      const props = { value: 1 };
      const state = RadioGroup.getDerivedStateFromProps(props);
      expect(state).toEqual({ value: 1 });
    });
    it('should return null if  value field not in props', () => {
      const props = {};
      const state = RadioGroup.getDerivedStateFromProps(props);
      expect(state).toBeNull();
    });
  });

  describe('#getValue', () => {
    it('should get initial state from props correctly if props has value field', () => {
      const wrapper = shallow(<RadioGroup value={1} />);
      expect(wrapper.state()).toEqual({ value: 1 });
    });
    it('should get initial state from props correctly if props has defaultValue field', () => {
      const wrapper = shallow(<RadioGroup defaultValue={1} />);
      expect(wrapper.state()).toEqual({ value: 1 });
    });

    it('should get null', () => {
      const wrapper = shallow(<RadioGroup />);
      expect(wrapper.state()).toEqual({ value: null });
    });
  });

  describe('#getChildChecked', () => {
    it('should get initial state from props correctly if component has single child', () => {
      const wrapper = shallow(
        <RadioGroup>
          <TestRadio checked value={233} />
        </RadioGroup>,
      );
      expect(wrapper.state()).toEqual({ value: 233 });
    });

    it('should get initial state from props correctly if component has children', () => {
      const wrapper = shallow(
        <RadioGroup>
          <TestRadio checked value={233} />
          <TestRadio checked value={222} />
        </RadioGroup>,
      );
      expect(wrapper.state()).toEqual({ value: 222 });
    });

    it('should get initial state from props correctly if component has invalid React element child', () => {
      const wrapper = shallow(
        <RadioGroup>
          <TestRadio checked={false} value={110} />
        </RadioGroup>,
      );
      expect(wrapper.state()).toEqual({ value: null });
    });
  });

  it('should render with children', () => {
    const wrapper = shallow(
      <RadioGroup id="test-radio-group">
        <TestRadio value={1} checked={false} />
        <TestRadio value={2} checked={false} />
      </RadioGroup>,
    );
    expect(wrapper.find('.za-radio-group__inner').children()).toHaveLength(2);
    expect(wrapper.prop('id')).toEqual('test-radio-group');
  });

  it('should render with cloned react element correctly', () => {
    const wrapper = shallow(
      <RadioGroup id="test-radio-group" type="button">
        <TestRadio value={1} />
        <TestRadio value={2} />
      </RadioGroup>,
    );
    expect(wrapper.find(TestRadio).at(0).props()).toEqual({
      value: 1,
      checked: false,
      type: 'button',
      shape: 'radius',
      disabled: false,
      onChange: expect.any(Function),
    });
    expect(wrapper.find(TestRadio).at(1).props()).toEqual({
      value: 2,
      checked: false,
      type: 'button',
      shape: 'radius',
      disabled: false,
      onChange: expect.any(Function),
    });
  });

  it('should handle onChange event on radio group component', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <RadioGroup id="test-radio-group" onChange={onChange}>
        <TestRadio value={1} checked={false} />
        <TestRadio value={2} checked={false} />
      </RadioGroup>,
    );
    expect(wrapper.state('value')).toBeNull();
    const input = wrapper.find('.za-radio-group__inner').children().at(0).find('input');
    expect(input.props()).toEqual(
      expect.objectContaining({
        value: 1,
        checked: false,
        type: undefined,
        shape: 'radius',
        disabled: false,
        onChange: expect.any(Function),
      }),
    );
    input.simulate('change');
    expect(onChange).toBeCalledWith(1);
    expect(wrapper.state('value')).toEqual(1);
  });

  it('should handle onChange event on the children', () => {
    const onChange1 = jest.fn();
    const onChange2 = jest.fn();
    const wrapper = mount(
      <RadioGroup id="test-radio-group">
        <TestRadio value={1} checked={false} onChange={onChange1} />
        <TestRadio value={2} checked={false} onChange={onChange2} />
      </RadioGroup>,
    );
    expect(wrapper.state('value')).toBeNull();
    const input = wrapper.find('.za-radio-group__inner').children().at(0).find('input');
    expect(input.props()).toEqual({
      value: 1,
      checked: false,
      type: undefined,
      shape: 'radius',
      disabled: false,
      onChange: expect.any(Function),
    });
    input.simulate('change');
    expect(onChange1).toBeCalledTimes(1);
    expect(onChange2).not.toBeCalled();
    expect(wrapper.state('value')).toEqual(1);
  });

  it('renders correctly', () => {
    const wrapper = render(
      <Radio.Group value="0" onChange={jest.fn()}>
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('receive new value', () => {
    const wrapper = shallow(
      <Radio.Group value="0" onChange={jest.fn()}>
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>,
    );
    wrapper.setProps({ value: '1' });
  });

  it('defaultValue', () => {
    const wrapper = render(
      <Radio.Group defaultValue="1">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Radio checked', () => {
    const wrapper = render(
      <Radio.Group>
        <Radio value="0">选项一</Radio>
        <Radio value="1" checked>
          选项二
        </Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // 圆角
  it('shape is radius', () => {
    const wrapper = shallow(
      <Radio.Group shape="radius">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>,
    );
    expect(wrapper.find('.za-radio-group').hasClass('za-radio-group--radius')).toBe(true);
  });

  // 椭圆角
  it('shape is round', () => {
    const wrapper = shallow(
      <Radio.Group shape="round">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>,
    );
    expect(wrapper.find('.za-radio-group').hasClass('za-radio-group--round')).toBe(true);
  });

  // 块级样式
  it('block', () => {
    const wrapper = shallow(
      <Radio.Group block>
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>,
    );
    expect(wrapper.find('.za-radio-group').hasClass('za-radio-group--block')).toBe(true);
  });

  // 列表样式
  it('type is cell', () => {
    const wrapper = shallow(
      <Radio.Group type="cell">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2" disabled>
          选项三
        </Radio>
      </Radio.Group>,
    );
    expect(wrapper.find('.za-radio-group').hasClass('za-radio-group--cell')).toBe(true);
  });

  it('radio group onChange event', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Radio.Group shape="round" onChange={onChange}>
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2" disabled>
          选项三
        </Radio>
      </Radio.Group>,
    );
    const firstCheckbox = wrapper.find(Radio).first().dive().find('input[type="radio"]');
    firstCheckbox.simulate('change', { target: { checked: true } });
    expect(onChange).toBeCalledWith('0');
  });
});
