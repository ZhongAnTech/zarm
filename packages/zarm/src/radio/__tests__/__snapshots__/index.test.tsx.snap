// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio defaultChecked 1`] = `
<span
  class="za-radio za-radio--checked"
>
  <span
    class="za-radio__widget"
  >
    <span
      class="za-radio__inner"
    />
  </span>
  <span
    class="za-radio__text"
  >
    选项一
  </span>
  <input
    checked=""
    class="za-radio__input"
    type="radio"
    value="0"
  />
</span>
`;

exports[`Radio renders correctly 1`] = `
<span
  class="za-radio"
>
  <span
    class="za-radio__widget"
  >
    <span
      class="za-radio__inner"
    />
  </span>
  <span
    class="za-radio__text"
  >
    选项一
  </span>
  <input
    class="za-radio__input"
    type="radio"
    value="0"
  />
</span>
`;

exports[`Radio type is button 1`] = `
<button
  class="za-radio"
  type="button"
>
  选项一
  <input
    class="za-radio__input"
    type="radio"
    value="0"
  />
</button>
`;

exports[`Radio type is cell 1`] = `
<div
  class="za-cell za-cell--link"
>
  <div
    class="za-cell__inner"
  >
    <div
      class="za-cell__header"
    />
    <div
      class="za-cell__body"
    >
      <div
        class="za-cell__content"
      >
        <span
          class="za-radio"
        >
          <span
            class="za-radio__widget"
          >
            <span
              class="za-radio__inner"
            />
          </span>
          <span
            class="za-radio__text"
          >
            选项一
          </span>
          <input
            class="za-radio__input"
            type="radio"
            value="0"
          />
        </span>
      </div>
    </div>
    <div
      class="za-cell__footer"
    />
  </div>
</div>
`;

exports[`Radio.Group Radio checked 1`] = `
<div
  class="za-radio-group za-radio-group--xs za-radio-group--radius"
>
  <div
    class="za-radio-group__inner"
  >
    <span
      class="za-radio"
    >
      <span
        class="za-radio__widget"
      >
        <span
          class="za-radio__inner"
        />
      </span>
      <span
        class="za-radio__text"
      >
        选项一
      </span>
      <input
        class="za-radio__input"
        type="radio"
        value="0"
      />
    </span>
    <span
      class="za-radio za-radio--checked"
    >
      <span
        class="za-radio__widget"
      >
        <span
          class="za-radio__inner"
        />
      </span>
      <span
        class="za-radio__text"
      >
        选项二
      </span>
      <input
        checked=""
        class="za-radio__input"
        type="radio"
        value="1"
      />
    </span>
    <span
      class="za-radio"
    >
      <span
        class="za-radio__widget"
      >
        <span
          class="za-radio__inner"
        />
      </span>
      <span
        class="za-radio__text"
      >
        选项三
      </span>
      <input
        class="za-radio__input"
        type="radio"
        value="2"
      />
    </span>
  </div>
</div>
`;

exports[`Radio.Group defaultValue 1`] = `
<div
  class="za-radio-group za-radio-group--xs za-radio-group--radius"
>
  <div
    class="za-radio-group__inner"
  >
    <span
      class="za-radio"
    >
      <span
        class="za-radio__widget"
      >
        <span
          class="za-radio__inner"
        />
      </span>
      <span
        class="za-radio__text"
      >
        选项一
      </span>
      <input
        class="za-radio__input"
        type="radio"
        value="0"
      />
    </span>
    <span
      class="za-radio za-radio--checked"
    >
      <span
        class="za-radio__widget"
      >
        <span
          class="za-radio__inner"
        />
      </span>
      <span
        class="za-radio__text"
      >
        选项二
      </span>
      <input
        checked=""
        class="za-radio__input"
        type="radio"
        value="1"
      />
    </span>
    <span
      class="za-radio"
    >
      <span
        class="za-radio__widget"
      >
        <span
          class="za-radio__inner"
        />
      </span>
      <span
        class="za-radio__text"
      >
        选项三
      </span>
      <input
        class="za-radio__input"
        type="radio"
        value="2"
      />
    </span>
  </div>
</div>
`;

exports[`Radio.Group renders correctly 1`] = `
<div
  class="za-radio-group za-radio-group--xs za-radio-group--radius"
>
  <div
    class="za-radio-group__inner"
  >
    <span
      class="za-radio za-radio--checked"
    >
      <span
        class="za-radio__widget"
      >
        <span
          class="za-radio__inner"
        />
      </span>
      <span
        class="za-radio__text"
      >
        选项一
      </span>
      <input
        checked=""
        class="za-radio__input"
        type="radio"
        value="0"
      />
    </span>
    <span
      class="za-radio"
    >
      <span
        class="za-radio__widget"
      >
        <span
          class="za-radio__inner"
        />
      </span>
      <span
        class="za-radio__text"
      >
        选项二
      </span>
      <input
        class="za-radio__input"
        type="radio"
        value="1"
      />
    </span>
    <span
      class="za-radio"
    >
      <span
        class="za-radio__widget"
      >
        <span
          class="za-radio__inner"
        />
      </span>
      <span
        class="za-radio__text"
      >
        选项三
      </span>
      <input
        class="za-radio__input"
        type="radio"
        value="2"
      />
    </span>
  </div>
</div>
`;
